File: leetcode/python/1239_maximum_length_of_concatenated_string.py
class Solution:
    def maxLength(self, arr: list[str]) -> int:

        def dfs(i: int, subseq: str) -> int:
            if len(subseq) != len(set(subseq)):
                return 0
            if i == len(arr):
                return len(subseq)

            return max(len(subseq), max([dfs(j+1, subseq+arr[j]) for j in range(i, len(arr))]))
        return dfs(0, '')

s = Solution()
print(s.maxLength(["un","iq","ue"]))
print(s.maxLength(["cha","r","act","ers"]))
print(s.maxLength(["jnfbyktlrqumowxd","mvhgcpxnjzrdei"]))

File: leetcode/rust/src/1239_maximum_length_of_concatenated_string.rs
use std::collections::HashSet;

impl Solution {
    pub fn max_length(arr: Vec<String>) -> i32 {
        fn dfs(arr: &Vec<String>, s: String, start: usize) -> i32 {
            let mut chars = s.chars().collect::<HashSet<_>>();
            if chars.len() != s.len() {
                return 0;
            }

            let mut res = s.len() as i32;
            for i in start..arr.len() {
                res = res.max(
                    dfs(arr, s.clone() + arr.get(i).unwrap(), i + 1));
            }
            res
        }
        dfs(&arr, "".to_string(), 0)
    }
}